fem/fe_ray_intersect.hpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2019, Lawrence Livermore National Security, LLC. Produced at
```

Comment:
Does this need to be in a separate file? Shouldn't it be a part of the `ElementTransfromation` routines? Will the algorithm sill work if the element transformation is not given by isoparametric finite elements (but we have its Jacobian, etc.)?

Reply:
Thanks @tzanio , I did take a look at the `ElementTransformation` routines, but, I wasn't sure if this type of operation would fit with the design of the methods therein. So, I added this in a separate file primarily because that seemed like the simplest approach and point of entry for me.

In this initial design, an application may use this as follows:
```
// set any parameters (optional, reasonable defaults are provided)
mfem::fe_ray_setMaxNewtonIteration( 20 );
// etc.
...
const FiniteElement* fe = mesh->GetNodalFESpace()->GetFE( ielem );
bool converged = mfem::fe_ray_solve( ielem, mesh, x0, n, r, t );
if ( converged && mfem::fe_ray_intersects( fe, r, t ) )
{
   // compute intersection point (assuming 3D)
   p[0]=x0[0]+n[0]*t;
   p[1]=x0[1]+n[1]*t;
   p[2]=x0[2]+n[2]*t;

    ....
}

...

```

How would you suggest to change the user-facing API if this becomes part of the `ElementTransformation` routines? 

From a style perspective, these routines do not hold any state information, so, it feels a bit clunky to require allocating an object to call a function instead of calling a function. Perhaps, they can be *static* methods of a class, that way they can be encapsulated within a namespace accordingly ?

The impetus for keeping a C-Style API more or less is motivated by the need to call these routines from within a kernel loop (projecting in to the future). We would typically have a bunch of rays whose intersection with a surface needs to be computed. Under this scenario, each thread would operate on a single ray checking a list of candidate elements for the intersection by calling `fe_ray_solve` and `fe_ray_intersects` respectively.

You make a good point about whether isoparametric elements are required or not. Given that we can evaluate a jacobian, as you mentioned, I would expect that this should work.  I'll add some tests using a different basis to verify.

Reply:
@tzanio I added a tests using `H1Pos_FiniteElementCollection` basis, in addition to the `H1_FECollection`, i.e., `Lagrange` basis that I was using before. However, I think those are still considered iso-parametric finite elements? Is that correct? 

Could you clarify what you meant by "element transformation not given by iso-parametric finite elements"? Is there a different `FiniteElementCollection` that we should have a test for?